pkg/s3select/csv/reader.go

```diff
@@ -17,157 +17,281 @@
+	close        chan struct{}    // used for shutting down the splitter before end of stream
+}
+
+// queueItem is an item in the queue.
+type queueItem struct {
+	input []byte          // raw input sent to the worker
+	dst   chan [][]string // result of block decode
+	err   error           // any error encountered will be set here
 }
 
 // Read - reads single record.
-func (r *Reader) Read() (sql.Record, error) {
-	csvRecord, err := r.csvReader.Read()
-	if err != nil {
-		if err != io.EOF {
-			return nil, errCSVParsingError(err)
+// Once Read is called the previous record should no longer be referenced.
+func (r *Reader) Read(dst sql.Record) (sql.Record, error) {
+	// If we have have any records left, return these before any error.
+	for len(r.current) <= r.recordsRead {
+		if r.err != nil {
+			return nil, r.err
 		}
-
-		return nil, err
+		// Move to next block
+		item, ok := <-r.queue
+		if !ok {
+			r.err = io.EOF
+			return nil, r.err
+		}
+		r.csvDstPool.Put(r.current)
+		r.current = <-item.dst
+		r.err = item.err
+		r.recordsRead = 0
 	}
+	csvRecord := r.current[r.recordsRead]
+	r.recordsRead++
 
+	// If no column names are set, use _(index)
 	if r.columnNames == nil {
 		r.columnNames = make([]string, len(csvRecord))
 		for i := range csvRecord {
 			r.columnNames[i] = fmt.Sprintf("_%v", i+1)
 		}
 	}
 
+	// If no index max, add that.
 	if r.nameIndexMap == nil {
 		r.nameIndexMap = make(map[string]int64)
 		for i := range r.columnNames {
 			r.nameIndexMap[r.columnNames[i]] = int64(i)
 		}
 	}
+	dstRec, ok := dst.(*Record)
+	if !ok {
+		dstRec = &Record{}
+	}
+	dstRec.columnNames = r.columnNames
+	dstRec.csvRecord = csvRecord
+	dstRec.nameIndexMap = r.nameIndexMap
 
-	return &Record{
-		columnNames:  r.columnNames,
-		csvRecord:    csvRecord,
-		nameIndexMap: r.nameIndexMap,
-	}, nil
+	return dstRec, nil
 }
 
-// Close - closes underlaying reader.
+// Close - closes underlying reader.
 func (r *Reader) Close() error {
+	if r.close != nil {
+		close(r.close)
+		r.close = nil
+	}
 	return r.readCloser.Close()
 }
 
-// NewReader - creates new CSV reader using readCloser.
-func NewReader(readCloser io.ReadCloser, args *ReaderArgs) (*Reader, error) {
-	if args == nil || args.IsEmpty() {
-		panic(fmt.Errorf("empty args passed %v", args))
+// nextSplit will attempt to skip a number of bytes and
+// return the buffer until the next newline occurs.
+// The last block will be sent along with an io.EOF.
+func (r *Reader) nextSplit(skip int, dst []byte) ([]byte, error) {
+	if cap(dst) < skip {
+		dst = make([]byte, 0, skip+1024)
+	}
+	dst = dst[:skip]
+	if skip > 0 {
+		n, err := io.ReadFull(r.buf, dst)
+		if err != nil && err != io.ErrUnexpectedEOF {
+			// If an EOF happens after reading some but not all the bytes,
+			// ReadFull returns ErrUnexpectedEOF.
+			return nil, err
+		}
+		dst = dst[:n]
 	}
+	// Read until next line.
+	in, err := r.buf.ReadBytes('\n')
+	dst = append(dst, in...)
```

Comment:
It is not guaranteed that a newline will be found here.

Reply:
Indeed, we could reach EOF. In that case we return the remaining bytes and io.EOF - ReadBytes states:
```
// If ReadBytes encounters an error before finding a delimiter,
// it returns the data read before the error and the error itself (often io.EOF).
```